src/main/java/org/links/objecthash/ObjectHash.java

```diff
@@ -0,0 +1,205 @@
+                                           + obj.getClass());
+      }
+    }
+
+  }
+
+  private void hashTaggedBytes(char tag, byte[] bytes)
+      throws NoSuchAlgorithmException {
+    digester.reset();
+    digester.update((byte) tag);
+    digester.update(bytes);
+    hash = digester.digest();
+  }
+
+  private void hashString(String str) throws NoSuchAlgorithmException {
+    hashTaggedBytes('u', str.getBytes());
+  }
+
+
+  private void hashList(JSONArray list) throws NoSuchAlgorithmException,
+                                               JSONException {
+    digester.reset();
+    digester.update((byte) ('l'));
+    for (int n = 0; n < list.length(); ++n) {
+      ObjectHash innerObject = new ObjectHash();
+      innerObject.hashAny(list.get(n));
+      digester.update(innerObject.hash());
+    }
+    hash = digester.digest();
+  }
+
+  private void hashObject(JSONObject obj) throws NoSuchAlgorithmException, 
+                                                 JSONException {
+    ByteBuffer buff = ByteBuffer.allocate(2 * obj.length() * SHA256_BLOCK_SIZE);
+    Iterator<String> keys = obj.keys();
+    while (keys.hasNext()) {
+      String key = keys.next();
+      // TODO(phad): would be nice to chain all these calls builder-stylee.
+      ObjectHash hKey = new ObjectHash();
+      hKey.hashString(key);
+      ObjectHash hVal = new ObjectHash();
+      hVal.hashAny(obj.get(key));
+      buff.put(hKey.hash());
+      buff.put(hVal.hash());
+    }
+    hashTaggedBytes('d', buff.array());
+  }
+
+  private static int parseHex(char digit) {
+    assert ((digit >= '0' && digit <= '9') || (digit >= 'a' && digit <= 'f'));
+    if (digit >= '0' && digit <= '9') {
+      return digit - '0';
+    } else {
+      return 10 + digit - 'a';
+    }
+  }
+
+  public static ObjectHash fromHex(String hex) throws NoSuchAlgorithmException {
+    ObjectHash h = new ObjectHash();
+    hex = hex.toLowerCase();
+    if (hex.length() % 2 == 1) {
+      hex = '0' + hex;
+    }
+    // TODO(phad): maybe just use Int.valueOf(s).intValue()
+    int pos = SHA256_BLOCK_SIZE;
+    for (int idx = hex.length(); idx > 0; idx -= 2) {
+      h.hash[--pos] = (byte) (16 * parseHex(hex.charAt(idx - 2))
+                              + parseHex(hex.charAt(idx - 1)));
+    }
+    return h;
+  }
+
+  private static JsonType getType(Object jsonObj) {
+    if (jsonObj instanceof JSONArray) {
+      return JsonType.ARRAY;
+    } else if (jsonObj instanceof JSONObject) {
+      return JsonType.OBJECT;
+    } else if (jsonObj instanceof String) {
+      return JsonType.STRING;
+    } else {
+      return JsonType.UNKNOWN;
+    }
+  }
+
+  public static ObjectHash commonJsonHash(String json)
+      throws JSONException, NoSuchAlgorithmException {
+    ObjectHash h = new ObjectHash();
+    JSONTokener tokener = new JSONTokener(json);
+    Object outerValue = tokener.nextValue();
+    JsonType outerType = getType(outerValue);
+    switch (outerType) {
```

Comment:
Because I assumed that [common|python]JsonHash will only be called with the complete JSON doc, and legal JSON only uses a list or dictionary at the top-most element.